sources/api/netdog/src/net_config/v3.rs

```diff
@@ -0,0 +1,119 @@
+//! The `v3` module contains the third version of the network configuration and implements the
+//! appropriate traits.
+
+use super::devices::NetworkDeviceV1;
+use super::{error, Interfaces, Result, Validate};
+use crate::interface_name::InterfaceName;
+use crate::wicked::{WickedInterface, WickedLinkConfig};
+use indexmap::IndexMap;
+use serde::Deserialize;
+use snafu::ensure;
+use std::collections::HashSet;
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct NetConfigV3 {
+    #[serde(flatten)]
+    pub(crate) net_devices: IndexMap<InterfaceName, NetworkDeviceV1>,
+}
+
+impl Interfaces for NetConfigV3 {
+    fn primary_interface(&self) -> Option<String> {
+        self.net_devices
+            .iter()
+            .find(|(_, v)| v.primary() == Some(true))
+            .or_else(|| self.net_devices.first())
+            .map(|(n, _)| n.to_string())
+    }
+
+    fn has_interfaces(&self) -> bool {
+        !self.net_devices.is_empty()
+    }
+
+    fn as_wicked_interfaces(&self) -> Vec<WickedInterface> {
+        let mut wicked_interfaces = Vec::new();
+        for (name, config) in &self.net_devices {
+            let interface = WickedInterface::from((name, config));
+
+            // If config is a Bond, we will generate the interface configuration for interfaces in
+            // that bond since we have all of the data and the bond consumes the device for other uses.
+            // For each interface: call WickedInterface::new(name), configure it and add that to
+            // wicked_interfaces Vec
+            if let NetworkDeviceV1::BondDevice(b) = config {
+                for device in &b.interfaces {
+                    let mut wicked_sub_interface = WickedInterface::new(device.clone());
+                    wicked_sub_interface.link = Some(WickedLinkConfig {
+                        master: name.clone(),
+                    });
+
+                    wicked_interfaces.push(wicked_sub_interface)
+                }
+            }
+
+            wicked_interfaces.push(interface)
+        }
+
+        wicked_interfaces
+    }
+}
+
+#[allow(clippy::to_string_in_format_args)]
```

Comment:
When you delete the `.to_string()` calls do you get some other clippy message? If so that must be a false positive (assuming the types implement `Display`).